### PR TITLE
Persisting to Neo4J without OGM

### PR DESCRIPTION
**Pull request by oxisto**
This PR is an attempt to build a lightweight persistence layer to Neo4J without the Neo4J OGM which causes immense problems (recursion depth, out of memory, etc.).

In the long run we might be able to get rid of the neo4j-ogm, which would allow us to re-design the node classes a little bit because we are not restricted by the OGM anymore.

- [x] Persist all `Edge` edges
- [x] Persist all AST nodes
- [x] Persist edge properties
- [x] Persist types + edges to nodes
- [x] Persist scopes + edges to nodes
- [x] Persist languages + edges to nodes

### CODE REVIEW

File: cpg-neo4j/src/main/kotlin/de/fraunhofer/aisec/cpg/persistence/Common.kt [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+val schemaPropertiesCache:
+    MutableMap<KClass<out Persistable>, Map<String, KProperty1<out Persistable, *>>> =
+    mutableMapOf()
+
+/**
+ * A set containing predefined property types represented as Kotlin type objects that can be used as
+ * properties in [schemaProperties].
+ */
+val propertyTypes =
+    setOf(
```
Comment (konradweiss) [inline cpg-neo4j/src/main/kotlin/de/fraunhofer/aisec/cpg/persistence/Common.kt:87]: Here we again have the issue that only hardcoded properties are persisted.

File: cpg-neo4j/src/main/kotlin/de/fraunhofer/aisec/cpg/persistence/Common.kt [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+ * @param property the property to be evaluated, belonging to a class implementing the Persistable
+ *   interface
+ * @return true if the property satisfies the conditions of being a "simple" property, false
+ *   otherwise
+ */
+private fun isSimpleProperty(property: KProperty1<out Persistable, *>): Boolean {
+    val returnType = property.returnType.withNullability(false)
+
+    return when {
+        property.visibility == KVisibility.PRIVATE -> false
```
Comment (konradweiss) [inline cpg-neo4j/src/main/kotlin/de/fraunhofer/aisec/cpg/persistence/Common.kt:236]: I am not sure if we want to have the visibility private to be a reason for not persisting a property and if it is a readoson it should definitely be its own function `isPrivate()`

File: cpg-neo4j/src/main/kotlin/de/fraunhofer/aisec/cpg/persistence/Common.kt (hunk 1 above)
Comment (konradweiss) [reply cpg-neo4j/src/main/kotlin/de/fraunhofer/aisec/cpg/persistence/Common.kt:87]: Is propertyType even used?

File: cpg-neo4j/src/main/kotlin/de/fraunhofer/aisec/cpg/persistence/Common.kt [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+ * [org.neo4j.ogm.typeconversion.AttributeConverter]. Since both of these classes are Neo4J OGM
+ * classes, we need to find new base types at some point.
+ */
+fun Any.convert(originalKey: String, properties: MutableMap<String, Any?>) {
+    // TODO: generalize conversions
+    if (this is Name && originalKey == "name") {
+        properties += NameConverter().toGraphProperties(this)
+    } else if (this is Name) {
+        properties.put(originalKey, SimpleNameConverter().toGraphProperty(this))
+    } else if (this is PhysicalLocation) {
```
Comment (konradweiss) [inline cpg-neo4j/src/main/kotlin/de/fraunhofer/aisec/cpg/persistence/Common.kt:130]: Time to integrate converter functionality into the converted class? `PhysicalLocation.toGraphProperties()`

**Review by konradweiss — CHANGES_REQUESTED** [review]
Some changes and discussion is still needed here. In some places I wonder if we can use introspectionfunctionality where we now use hardcoded propertyTypes or relationships

**Comment by KuechA** [comment]
@oxisto can you add a test that #234 is fixed?

**Review by konradweiss — APPROVED** [review]

File: cpg-neo4j/src/main/kotlin/de/fraunhofer/aisec/cpg/persistence/Common.kt (hunk 2 above)
Comment (oxisto) [reply cpg-neo4j/src/main/kotlin/de/fraunhofer/aisec/cpg/persistence/Common.kt]: We can't access private properties without major reflection foo, that's why I chose to ignore them.

File: cpg-neo4j/src/main/kotlin/de/fraunhofer/aisec/cpg/persistence/Common.kt (hunk 1 above)
Comment (oxisto) [reply cpg-neo4j/src/main/kotlin/de/fraunhofer/aisec/cpg/persistence/Common.kt]: Removed

File: cpg-neo4j/src/main/kotlin/de/fraunhofer/aisec/cpg/persistence/Neo4J.kt [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+fun TranslationResult.persist() {
+    val b = Benchmark(Persistable::class.java, "Persisting translation result")
+
+    val astNodes = this@persist.nodes
+    val scopes = this.finalCtx.scopeManager.filterScopes { true }
+    val languages = this.finalCtx.config.languages
+    val types =
+        (this.finalCtx.typeManager.firstOrderTypes + this.finalCtx.typeManager.secondOrderTypes)
+            .toIdentitySet()
+    val nodes = listOf(astNodes, scopes, languages, types).flatten()
```
[the remark this answers is not in the document]
Comment (oxisto) [reply cpg-neo4j/src/main/kotlin/de/fraunhofer/aisec/cpg/persistence/Neo4J.kt]: Fixed

File: cpg-neo4j/src/main/kotlin/de/fraunhofer/aisec/cpg/persistence/Neo4J.kt [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+        )
+    }
+}
+
+/**
+ * Some of our relationships are not real "edges" (i.e., [Edge]) (yet). We need to handle these case
+ * separately (for now).
+ */
+context(Session)
+private fun List<Node>.persistExtraRelationships() {
```
[the remark this answers is not in the document]
Comment (oxisto) [reply cpg-neo4j/src/main/kotlin/de/fraunhofer/aisec/cpg/persistence/Neo4J.kt]: Not anymore. Fixed, all relationships to nodes, even though they are not "edges" (yet) are persisted.

**Comment by oxisto** [comment]
> @oxisto can you add a test that #234 is fixed?

Done

**Pull request merged**